src/PAModel/Utility/Utilities.cs

```diff
@@ -358,5 +360,41 @@ public static string GetResourceRelativePath(ContentKind contentType)
                     throw new NotSupportedException("Unrecognized Content Kind for local resource");
             }
         }
+
+        /// <summary>
+        /// If the name length is longer than 50, it is truncated and appended with a hash (to avoid collisions).
+        /// Checks the length of the escaped name, since its possible that the length is under 60 before escaping but goes beyond 60 later.
+        /// We do modulo by 1000 of the hash to limit it to 3 characters.
+        /// </summary>
+        /// <param name="name">The name that needs to be truncated.</param>
+        /// <returns>Returns the truncated name if the escaped version of it is longer than 50 characters.</returns>
+        public static string TruncateNameIfTooLong(string name)
+        {
+            var escapedName = EscapeFilename(name);
+            if (escapedName.Length > MaxNameLength)
+            {
+                // limit the hash to 3 characters by doing a module by 4096 (16^3)
+                var hash = (GetHash(escapedName) % 4096).ToString("x3");
+                escapedName = escapedName.Substring(0, MaxNameLength - hash.Length) + hash;
```

Comment:
>hash [](http://example.com/codeflow?start=86&length=4)

nit:
Can we inject an underscore, makes it easier to realize it's been truncated and hashed.  Like:

+ "_" + hash;

"FileNameTruncated_fff"